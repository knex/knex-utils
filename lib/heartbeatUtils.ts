import Knex from 'knex'

export const HEARTBEAT_QUERIES = Object.freeze({
  ORACLE: 'select 1 from DUAL',
  POSTGRESQL: 'SELECT 1',
  MYSQL: 'SELECT 1',
  MSSQL: 'SELECT 1',
  SQLITE: 'SELECT 1',
  DEFAULT: 'SELECT 1',
})

export interface HeartbeatResult {
  isOk: boolean
  error?: Error
}

/**
 *
 * @param {Object} knex - Knex instance
 * @param {string} heartbeatQuery - SQL query that will be executed to check if connection is valid
 * @returns Promise<{Object|undefined}> wrapped error if connection is not valid, wrapped 'isOk: true' if it is valid, undefined if connection does not yet exist
 */
export function checkHeartbeat(
  knex: Knex,
  heartbeatQuery = HEARTBEAT_QUERIES.DEFAULT
): Promise<HeartbeatResult> {
  if (!knex) {
    throw new Error('Knex is mandatory parameter')
  }

  return knex
    .raw(heartbeatQuery)
    .then(() => {
      return {
        isOk: true,
      }
    })
    .catch((err: Error) => {
      return {
        isOk: false,
        error: err,
      }
    })
}

export type returnCallback<T> = (err?: any, result?: T) => any

export type CoinpaymentsRatesOpts = any
//export type returnCallback<T> = any
export type CoinpaymentsRatesResponse = any

export type CMDS = any

export class Something {
  public mapPayload<t>(x: any, y: any): any

  public rates(
    options?: CoinpaymentsRatesOpts | returnCallback<CoinpaymentsRatesResponse>,
    callback?: returnCallback
  ) {
    if (!options && !callback) {
      options = {}
    }
    if (typeof options === 'function') {
      callback = options
      options = {}
    }
    const requestPayload = mapPayload<CoinpaymentsRatesOpts>(options, {
      cmd: CMDS.RATES,
    })

    return {}
  }
}
