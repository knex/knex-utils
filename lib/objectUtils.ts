export function copyWithoutUndefined<T extends Record<K, V>, K extends string | number | symbol, V>(
  originalValue: T
): T {
  return Object.keys(originalValue).reduce((acc, key) => {
    // @ts-ignore
    if (originalValue[key] !== undefined) {
      // @ts-ignore
      acc[key] = originalValue[key]
    }
    return acc
  }, {} as Record<string, any>) as T
}

export function pick<T, K extends string | number | symbol>(
  source: T,
  names: readonly K[]
): Pick<T, Exclude<keyof T, Exclude<keyof T, K>>> {
  const result = {} as T
  let idx = 0
  while (idx < names.length) {
    if (names[idx] in source) {
      // @ts-ignore
      result[names[idx]] = source[names[idx]]
    }
    idx += 1
  }
  return result
}
