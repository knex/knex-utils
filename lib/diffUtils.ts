import { Knex } from 'knex'
import { pick } from './objectUtils'
import { chunk } from './chunkUtils'

function arrayIncludesWith(array: any[], target: any, comparator: any) {
  if (array == null) {
    return false
  }

  for (const value of array) {
    if (comparator(target, value)) {
      return true
    }
  }
  return false
}

function baseDifference<T>(initialArray: T[], newArray: T[], comparator: any): EntityListDiff<T> {
  const includes = arrayIncludesWith
  const newEntries: T[] = []
  const removedEntries: T[] = []

  for (const value of initialArray) {
    if (!includes(newArray, value, comparator)) {
      removedEntries.push(value)
    }
  }

  for (const value of newArray) {
    if (!includes(initialArray, value, comparator)) {
      newEntries.push(value)
    }
  }

  return {
    removedEntries,
    newEntries,
  }
}

export type EntityListDiff<T> = {
  newEntries: T[]
  removedEntries: T[]
}

export function calculateEntityListDiff<T>(
  oldList: T[],
  newList: T[],
  idFields: string[]
): EntityListDiff<T> {
  const comparator = (value1: any, value2: any) => {
    for (const idField of idFields) {
      if (value1[idField] !== value2[idField]) {
        return false
      }
    }
    return true
  }

  return baseDifference(oldList, newList, comparator)
}

export type UpdateJoinTableParams = {
  filterCriteria: Record<string, any>
  table: string
  primaryKeyField?: string
  chunkSize?: number
  idFields: string[]
  transactionProvider?: Knex.TransactionProvider
  transaction?: Knex.Transaction
}

async function getKnexOrTrx(
  knex: Knex,
  params: UpdateJoinTableParams
): Promise<Knex | Knex.Transaction> {
  if (knex.client.driverName === 'sqlite3') {
    return knex
  }

  if (params.transaction) {
    return params.transaction
  }

  if (params.transactionProvider) {
    return params.transactionProvider()
  }

  return knex.transaction()
}

export async function updateJoinTable<T>(
  knex: Knex,
  newList: T[],
  params: UpdateJoinTableParams
): Promise<EntityListDiff<T>> {
  const chunkSize = params.chunkSize || 100
  const trx = await getKnexOrTrx(knex, params)
  try {
    const oldList = await trx(params.table).select('*').where(params.filterCriteria)
    const diff = calculateEntityListDiff(oldList, newList, params.idFields)

    const insertChunks = chunk(diff.newEntries, chunkSize)
    for (const insertChunk of insertChunks) {
      await trx(params.table).insert(insertChunk)
    }

    // If we have a primary key, then we can delete in batch
    if (params.primaryKeyField) {
      const deleteIds = diff.removedEntries.map((entry) => {
        return entry[params.primaryKeyField!]
      })
      const deleteChunks = chunk(deleteIds, chunkSize)
      for (const deleteChunk of deleteChunks) {
        await trx(params.table).delete().whereIn(params.primaryKeyField, deleteChunk)
      }
    }
    // Otherwise we have to delete one-by-one
    else {
      const deleteCriteria = diff.removedEntries.map((entry) => {
        return pick(entry, params.idFields)
      })
      for (const entry of deleteCriteria) {
        await trx(params.table).delete().where(entry)
      }
    }

    if (trx.isTransaction && !params.transaction && !params.transactionProvider) {
      await (trx as Knex.Transaction).commit()
    }
    return diff
  } catch (err) {
    if (trx.isTransaction) {
      await (trx as Knex.Transaction).rollback()
    }
    throw err
  }
}
