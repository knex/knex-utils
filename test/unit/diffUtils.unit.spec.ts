import { calculateEntityListDiff } from '../../lib/diffUtils'

describe('diffUtils', () => {
  describe('calculateEntityListDiff (single field)', () => {
    it('returns empty result for identical arrays', () => {
      const array1 = [{ id: 1 }, { id: 2 }]
      const array2 = [{ id: 2 }, { id: 1 }]

      const result = calculateEntityListDiff(array1, array2, ['id'])

      expect(result).toMatchSnapshot()
    })

    it('returns correct result for new entries', () => {
      const array1 = [{ id: 1 }, { id: 2 }]
      const array2 = [{ id: 2 }, { id: 1 }, { id: 3 }]

      const result = calculateEntityListDiff(array1, array2, ['id'])

      expect(result).toMatchSnapshot()
    })

    it('returns correct result for removed entries', () => {
      const array1 = [{ id: 1 }, { id: 2 }, { id: 3 }]
      const array2 = [{ id: 2 }, { id: 1 }]

      const result = calculateEntityListDiff(array1, array2, ['id'])

      expect(result).toMatchSnapshot()
    })

    it('returns correct result for both added and removed entries', () => {
      const array1 = [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }]
      const array2 = [{ id: 2 }, { id: 1 }, { id: 5 }, { id: 6 }]

      const result = calculateEntityListDiff(array1, array2, ['id'])

      expect(result).toMatchSnapshot()
    })
  })

  describe('calculateEntityListDiff', () => {
    it('returns empty result for identical arrays', () => {
      const array1 = [
        { userId: 1, orgId: 1 },
        { userId: 2, orgId: 1 },
        { userId: 3, orgId: 2 },
      ]
      const array2 = [
        { userId: 1, orgId: 1 },
        { userId: 2, orgId: 1 },
        { userId: 3, orgId: 2 },
      ]

      const result = calculateEntityListDiff(array1, array2, ['userId', 'orgId'])

      expect(result).toMatchSnapshot()
    })

    it('returns correct result for new entries', () => {
      const array1 = [
        { userId: 1, orgId: 1 },
        { userId: 2, orgId: 1 },
      ]
      const array2 = [
        { userId: 1, orgId: 1 },
        { userId: 2, orgId: 1 },
        { userId: 3, orgId: 2 },
      ]

      const result = calculateEntityListDiff(array1, array2, ['userId', 'orgId'])

      expect(result).toMatchSnapshot()
    })

    it('returns correct result for removed entries', () => {
      const array1 = [
        { userId: 1, orgId: 1 },
        { userId: 2, orgId: 1 },
        { userId: 3, orgId: 2 },
      ]
      const array2 = [
        { userId: 1, orgId: 1 },
        { userId: 2, orgId: 1 },
      ]

      const result = calculateEntityListDiff(array1, array2, ['userId', 'orgId'])

      expect(result).toMatchSnapshot()
    })

    it('returns correct result for both added and removed entries', () => {
      const array1 = [
        { userId: 1, orgId: 1 },
        { userId: 2, orgId: 1 },
      ]
      const array2 = [
        { userId: 2, orgId: 1 },
        { userId: 3, orgId: 2 },
      ]

      const result = calculateEntityListDiff(array1, array2, ['userId', 'orgId'])

      expect(result).toMatchSnapshot()
    })
  })
})
