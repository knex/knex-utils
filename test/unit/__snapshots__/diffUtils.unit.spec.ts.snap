// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diffUtils calculateEntityDiff returns correct result for both added and removed entries 1`] = `
Object {
  "newEntries": Array [
    Object {
      "id": 5,
    },
    Object {
      "id": 6,
    },
  ],
  "removedEntries": Array [
    Object {
      "id": 3,
    },
    Object {
      "id": 4,
    },
  ],
}
`;

exports[`diffUtils calculateEntityDiff returns correct result for new entries 1`] = `
Object {
  "newEntries": Array [
    Object {
      "id": 3,
    },
  ],
  "removedEntries": Array [],
}
`;

exports[`diffUtils calculateEntityDiff returns correct result for removed entries 1`] = `
Object {
  "newEntries": Array [],
  "removedEntries": Array [
    Object {
      "id": 3,
    },
  ],
}
`;

exports[`diffUtils calculateEntityDiff returns empty result for identical arrays 1`] = `
Object {
  "newEntries": Array [],
  "removedEntries": Array [],
}
`;

exports[`diffUtils calculateJoinTableDiff returns correct result for both added and removed entries 1`] = `
Object {
  "newEntries": Array [
    Object {
      "orgId": 2,
      "userId": 3,
    },
  ],
  "removedEntries": Array [
    Object {
      "orgId": 1,
      "userId": 1,
    },
  ],
}
`;

exports[`diffUtils calculateJoinTableDiff returns correct result for new entries 1`] = `
Object {
  "newEntries": Array [
    Object {
      "orgId": 2,
      "userId": 3,
    },
  ],
  "removedEntries": Array [],
}
`;

exports[`diffUtils calculateJoinTableDiff returns correct result for removed entries 1`] = `
Object {
  "newEntries": Array [],
  "removedEntries": Array [
    Object {
      "orgId": 2,
      "userId": 3,
    },
  ],
}
`;

exports[`diffUtils calculateJoinTableDiff returns empty result for identical arrays 1`] = `
Object {
  "newEntries": Array [],
  "removedEntries": Array [],
}
`;
